apps/aecontract/test/aecontract_SUITE.erl

```diff
@@ -730,6 +750,170 @@ sophia_oracles(_Cfg) ->
+                                         #{ height => Height, amount => QFee }),
+            AbsTTLq = ttl_height(Height, TTLq),
+            AbsTTLr = ttl_height(AbsTTLq, TTLr),
+            %% Latest possible response expiry must be before oracle expiry and
+            %% response TTL must be relative.
+            case TTLo >= AbsTTLr andalso element(1, TTLr) == delta of
+                true ->
+                    true = is_integer(Res),
+                    St#{ query => Res, query_ttl => AbsTTLq, reply_ttl => TTLr };
+                _ -> Error = Res, St
+            end;
+        respond ->
+            #{ oracle := Oracle, query := Query, query_ttl := TTLq, reply_ttl := TTLr } = St,
+            Res = ?call(call_contract, Acc, Ct, respond, {tuple, []}, {Oracle, Query, 0, 42}, #{ height => Height }),
+            case TTLq >= Height of  %% Query must not have expired
+                true  -> St#{ reply_ttl => ttl_height(Height, TTLr) };
+                false -> Error = Res, St
+            end;
+        getAnswer ->
+            #{ oracle := Oracle, query := Query, reply_ttl := TTLr } = St,
+            ExpectedRes = case TTLr >= Height of   %% Response must not have expired,
+                            true  -> {some, 42};   %% and we only ask after responding.
+                            false -> none
+                          end,
+            ExpectedRes = ?call(call_contract, Acc, Ct, getAnswer, {option, word}, {Oracle, Query}, #{ height => Height }),
+            St
+    end.
+
+ttls(Now, Ts) ->
+    [{block, T} || T <- Ts] ++
+    [{delta, T - Now} || T <- Ts, T > Now].
+
+%% Base scenario for setting up and possibly extending an oracle.
+ttl_scenario_create_and_extend(Start0, Extend0, Stop0) ->
+    List = fun(X) when is_list(X) -> X; (X) -> [X] end,
+    [ [ {Start,  {create, TTLo}} ] ++
+      [ {Extend, {extend, TTLe}} || Extend /= false ]
+      || Start <- List(Start0), Extend <- List(Extend0), Stop <- List(Stop0),
+         TTLo <- [ T || Extend == false, T <- ttls(Start, [Stop]) ] ++
+                 [ T || Extend /= false, T <- ttls(Start, [Extend + 5]) ],
+         TTLe <- [ 0 || Extend == false ] ++
+                 [ T || Extend /= false, T <- ttls(ttl_height(Start, TTLo), [Stop]) ] ].
+
+%% Base scenario for setting up a query
+ttl_scenario_create_query(Start0, Extend0, Query0, QTTL0, RTTL0, Stop0) ->
+    List = fun(X) when is_list(X) -> X; (X) -> [X] end,
+    [ combine_ttl_scenarios(Setup, [{Query, {query, QTTL, {delta, RTTL}}}])
+    || Setup <- ttl_scenario_create_and_extend(Start0, Extend0, Stop0),
+       Query <- List(Query0),
+       QTTL  <- ttls(Query, List(QTTL0)),
+       RTTL  <- List(RTTL0) ].
+
+%% Extending an oracle must be done before expiry.
+ttl_scenario_extend_after_expiry() ->
+    [ [ {10, {create, TTLo}},
+        {20, {extend, TTLe}} ]
+    || TTLo <- ttls(10, [15]),
+       TTLe <- ttls(20, [25]) ].
```

Comment:
If I read this correctly, the following cases are not tested:
* Oracle extend VM primop called with absolute expiry TTL strictly smaller than current height;
* Oracle extend VM primop called with absolute expiry TTL equal to current height;
* Oracle extend VM primop called with absolute expiry TTL greater than current height though strictly smaller than planned expiration of oracle;
* Oracle extend VM primop called with absolute expiry TTL greater than current height though equal to planned expiration of oracle.

Reply:
I've added tests for this (0adce03).